dcos/util.go

```diff
@@ -0,0 +1,297 @@
+	}
+
+	validIP := net.ParseIP(detectedIP)
+	if validIP == nil {
+		return nil, ErrDCOSUtil{fmt.Sprintf("command %s return invalid IP address %s", FileDetectIP, detectedIP)}
+	}
+
+	return validIP, nil
+}
+
+// Role return a node role.
+func (d *dcosUtil) Role() (string, error) {
+	var isMaster, isAgent, isAgentPublic bool
+	isFile := func(path string, flag *bool) {
+		if _, err := os.Stat(path); err == nil {
+			*flag = true
+		}
+	}
+
+	// iterate over files and get the role.
+	for _, i := range []struct {
+		file string
+		flag *bool
+	}{
+		{
+			file: d.roleMasterFile,
+			flag: &isMaster,
+		},
+		{
+			file: d.roleAgentFile,
+			flag: &isAgent,
+		},
+		{
+			file: d.roleAgentPublicFile,
+			flag: &isAgentPublic,
+		},
+	} {
+		isFile(i.file, i.flag)
+	}
+
+	// make sure we have only one role
+	if isMaster && (isAgent || isAgentPublic) {
+		return "", ErrDCOSUtil{"Node cannot be master and agent / agent_public at the same time"}
+	}
+
+	if isAgent && isAgentPublic {
+		return "", ErrDCOSUtil{"Node cannot be agent and agent_public at the same time"}
+	}
+
+	if !isMaster && !isAgent && !isAgentPublic {
+		return "", ErrDCOSUtil{"Unable to determine node role"}
+	}
+
+	// iterate over roles
+	for _, nodeRole := range []struct {
+		flag bool
+		role string
+	}{
+		{
+			flag: isMaster,
+			role: RoleMaster,
+		},
+		{
+			flag: isAgent,
+			role: RoleAgent,
+		},
+		{
+			flag: isAgentPublic,
+			role: RoleAgentPublic,
+		},
+	} {
+		if nodeRole.flag {
+			return nodeRole.role, nil
+		}
+	}
+
+	return "", ErrDCOSUtil{"Unable to determine node role"}
+}
+
+// IsLeader checks if the node is leader.
+func (d *dcosUtil) IsLeader() (bool, error) {
+	role, err := d.Role()
+	if err != nil {
+		return false, err
+	}
+
+	// agent cannot be leader
+	if role != RoleMaster {
+		return false, nil
+	}
+
+	addrs, err := net.LookupIP(DNSRecordLeader)
+	if err != nil {
+		return false, err
+	}
+
+	if len(addrs) != 1 {
+		return false, ErrDCOSUtil{fmt.Sprintf("Error getting mesos leader. Number of ip addresses %d", len(addrs))}
+	}
+
+	localIP, err := d.DetectIP()
+	if err != nil {
+		return false, err
+	}
+
+	if localIP.Equal(addrs[0]) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// MesosID returns a mesosID for leading master and agents.
+func (d *dcosUtil) MesosID() (string, error) {
+	if d.client == nil {
+		return "", ErrDCOSUtil{"Unable to get mesos ID. Uninitialized http client"}
+	}
+
+	if d.mesosStateURL == "" {
+		return "", ErrDCOSUtil{"Unable to get mesos ID. Uninitialized url"}
+	}
+
+	req, err := http.NewRequest("GET", d.mesosStateURL, nil)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", ErrDCOSUtil{fmt.Sprintf("GET request to %s returned response code %d", d.mesosStateURL, resp.StatusCode)}
+	}
+
+	type stateJSON struct {
```

Comment:
too bad you need to pull back the whole state json. would be really nice if there was another endpoint you could hit to give you the slave info. what about `/slaves`? https://github.com/apache/mesos/blob/master/docs/endpoints/master/slaves.md